File: decoders/decoder.js
const crypto = require('crypto');
const axios = require('axios');

const APIKEY = process.env.APIKEY;

async function transitDecrypt(ciphertext, keyLength = 32) {
    const epoch = Math.floor(Date.now() / 60000);
    const hash = crypto.createHash('sha256');
    hash.update(`${epoch}.${APIKEY}`);
    const aesKey = hash.digest().slice(0, keyLength);

    const bufferCiphertext = Buffer.from(ciphertext, 'base64');
    if (bufferCiphertext.length < 12 + 16) {
        throw new Error('Ciphertext is too short');
    }

    const iv = bufferCiphertext.slice(0, 12); // First 12 bytes
    const authTag = bufferCiphertext.slice(bufferCiphertext.length - 16); // Last 16 bytes
    const encryptedData = bufferCiphertext.slice(12, bufferCiphertext.length - 16); // Data in between

    const decipher = crypto.createDecipheriv('aes-256-gcm', aesKey, iv);
    decipher.setAuthTag(authTag); // Set the authentication tag

    let decrypted;
    try {
        decrypted = Buffer.concat([decipher.update(encryptedData), decipher.final()]);
    } catch (err) {
        throw new Error('Decryption failed: ' + err.message);
    }

    return JSON.parse(decrypted.toString());
}

async function getCipher() {
    const headers = {
        'accept': 'application/json',
        'Authorization': `Bearer ${APIKEY}`,
    };
    const params = {
        table_name: 'default',
    };
    const response = await axios.get('http://0.0.0.0:8080/get-table', {params, headers});
    if (response.status !== 200) {
        throw new Error(response.data);
    }
    return response.data.detail;
}

async function main() {
    try {
        const ciphertext = await getCipher();
        const result = await transitDecrypt(ciphertext);
        console.log(result);
    } catch (error) {
        console.error(error);
    }
}

main();

File: decoders/decoder.py
import base64
import hashlib
import json
import os
import time
from typing import Dict, ByteString, Any

import requests
from cryptography.hazmat.primitives.ciphers.aead import AESGCM

APIKEY = os.environ["APIKEY"]


def transit_decrypt(ciphertext: str | ByteString, key_length: int = 32) -> Dict[str, Any]:
    epoch = int(time.time()) // 60
    hash_object = hashlib.sha256(f"{epoch}.{APIKEY}".encode())
    aes_key = hash_object.digest()[:key_length]
    if isinstance(ciphertext, str):
        ciphertext = base64.b64decode(ciphertext)
    decrypted = AESGCM(aes_key).decrypt(ciphertext[:12], ciphertext[12:], b"")
    return json.loads(decrypted)


def get_cipher():
    headers = {
        'accept': 'application/json',
        'Authorization': f'Bearer {APIKEY}',
    }
    params = {
        'table_name': 'default',
    }
    response = requests.get('http://0.0.0.0:8080/get-table', params=params, headers=headers)
    assert response.ok, response.text
    return response.json()['detail']


print(transit_decrypt(ciphertext=get_cipher()))
